### PR TITLE
Builder matrix determinant and inversion

### PR DESCRIPTION
**Pull request by trenouf**
This adds matrix determinant and inversion operations to Builder, and
switches the SPIR-V reader to use them.

Change-Id: I9e41975ad0814bab687bff684ad2814d7c87e1e8

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit a7c50ebf7be81fe451503f10eeb74ce9f41d2037
<details open><summary>Driver commits used in build</summary><ul>

<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 66e78b997748d03d77e1d706c10f1f17e18e5654</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 9b632ef4f132bddc94769702ed8b49efbc39d89c</li>
<li>LLVM: amd-vulkan-dev 9bc5dd4450a6361faf5c5661056a7ee494fad830</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 282837/282837 (100.0%)</li>
<li>Failed: 0/282837 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366376/366376 (100.0%)</li>
<li>Failed: 0/366376 (0.0%)</li>
</ul></details>


</ul></details>

File: builder/llpcBuilder.h [hunk 1]
```diff
@@ -802,6 +802,17 @@ class Builder : public IRBuilder<>
...
         const Twine& instName = "") = 0;  // [in] Name to give instruction(s)
 
+    // Create matrix determinant operation. Matrix must be square
+    virtual Value* CreateDeterminant(
+        Value* const pMatrix,             // [in] Matrix
+        const Twine& instName = "") = 0;  // [in] Name to give instruction(s)
+
+    // Create matrix inverse operation. Matrix must be square. Result is undefined if the matrix
+    // is singular or poorly conditioned (nearly singular).
+    virtual Value* CreateInvertMatrix(
```
Comment (amdrexu) [inline builder/llpcBuilder.h]: CreateInvertMatrix -> CreateMatrixInverse
Align it with GLSL.std.450 spec

File: builder/llpcBuilderImpl.h [hunk 2]
```diff
@@ -472,9 +472,25 @@ class BuilderImplMatrix : virtual public BuilderImplBase
                               Value* const pVector2,
                               const Twine& instName = "") override final;
 
+    // Create matrix determinant operation.
+    Value* CreateDeterminant(Value* const pMatrix, const Twine& instName = "") override final;
+
+    // Create matrix inverse operation.
+    Value* CreateInvertMatrix(Value* const pMatrix, const Twine& instName = "") override final;
```
Comment (amdrexu) [inline builder/llpcBuilderImpl.h]: CreateInvertMatrix -> CreateMatrixInverse

File: builder/llpcBuilderImplMatrix.cpp [hunk 3]
```diff
@@ -203,5 +209,180 @@ Value* BuilderImplMatrix::CreateOuterProduct(
...
+            }
+            ++inIdx;
+        }
+    }
+}
+
+// =====================================================================================================================
+// Create matrix inverse operation. Matrix must be square. Result is undefined if the matrix
+// is singular or poorly conditioned (nearly singular).
+Value* BuilderImplMatrix::CreateInvertMatrix(
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: CreateInvertMatrix -> CreateMatrixInverse

File: builder/llpcBuilderRecorder.cpp [hunk 4]
```diff
@@ -93,6 +93,10 @@ StringRef BuilderRecorder::GetCallName(
         return "matrix.times.matrix";
     case Opcode::OuterProduct:
         return "outer.product";
+    case Opcode::Determinant:
+        return "determinant";
+    case Opcode::InvertMatrix:
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: InvertMatrix -> MatrixInverse
matrix.inverse

File: builder/llpcBuilderRecorder.cpp [hunk 5]
```diff
@@ -350,6 +354,27 @@ Value* BuilderRecorder::CreateOuterProduct(
...
+{
+    return Record(Determinant,
+                  pMatrix->getType()->getArrayElementType()->getVectorElementType(),
+                  pMatrix,
+                  instName);
+}
+
+// =====================================================================================================================
+// Create calculation of matrix inverse
+Value* BuilderRecorder::CreateInvertMatrix(
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: InvertMatrix -> MatrixInverse

File: builder/llpcBuilderRecorder.cpp [hunk 6]
```diff
@@ -350,6 +354,27 @@ Value* BuilderRecorder::CreateOuterProduct(
     return Record(Opcode::OuterProduct, pResultTy, { pVector1, pVector2 }, instName);
 }
 
+// =====================================================================================================================
+// Create calculation of matrix determinant
+Value* BuilderRecorder::CreateDeterminant(
+        Value* const pMatrix,            // [in] Matrix
+        const Twine& instName)
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: Add the missing comments:
// [in] Name to give instruction(s)

File: builder/llpcBuilderRecorder.cpp [hunk 7]
```diff
@@ -350,6 +354,27 @@ Value* BuilderRecorder::CreateOuterProduct(
...
+                  pMatrix->getType()->getArrayElementType()->getVectorElementType(),
+                  pMatrix,
+                  instName);
+}
+
+// =====================================================================================================================
+// Create calculation of matrix inverse
+Value* BuilderRecorder::CreateInvertMatrix(
+        Value* const pMatrix,            // [in] Matrix
+        const Twine& instName)
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: Add the missing comments:
// [in] Name to give instruction(s)

File: builder/llpcBuilderImpl.h [hunk 8]
```diff
@@ -472,9 +472,25 @@ class BuilderImplMatrix : virtual public BuilderImplBase
...
+
 private:
     LLPC_DISALLOW_DEFAULT_CTOR(BuilderImplMatrix)
     LLPC_DISALLOW_COPY_AND_ASSIGN(BuilderImplMatrix)
+
+    // Helper function for determinant calculation
+    Value* Determinant(ArrayRef<Value*> elements, uint32_t size);
```
Comment (amdrexu) [inline builder/llpcBuilderImpl.h]: size -> order

File: builder/llpcBuilderImplMatrix.cpp [hunk 9]
```diff
@@ -203,5 +209,180 @@ Value* BuilderImplMatrix::CreateOuterProduct(
...
+    Value* pResult = Determinant(elements, size);
+    pResult->setName(instName);
+    return pResult;
+}
+
+// =====================================================================================================================
+// Helper function for determinant calculation
+Value* BuilderImplMatrix::Determinant(
+    ArrayRef<Value*>    elements,   // Elements of matrix
+    uint32_t            size)       // Number of rows and columns in matrix
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: size -> order

File: builder/llpcBuilderImplMatrix.cpp [hunk 10]
```diff
@@ -203,5 +209,180 @@ Value* BuilderImplMatrix::CreateOuterProduct(
...
     return pResult;
 }
+
+// =====================================================================================================================
+// Create matrix determinant operation. Matrix must be square
+Value* BuilderImplMatrix::CreateDeterminant(
+    Value* const pMatrix,     // [in] Matrix
+    const Twine& instName)    // [in] Name to give instruction(s)
+{
+    uint32_t size = pMatrix->getType()->getArrayNumElements();
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: size  -> order

File: builder/llpcBuilderImplMatrix.cpp [hunk 11]
```diff
@@ -203,5 +209,180 @@ Value* BuilderImplMatrix::CreateOuterProduct(
...
+// =====================================================================================================================
+// Get submatrix by deleting specified row and column
+void BuilderImplMatrix::GetSubmatrix(
+    ArrayRef<Value*>        matrix,         // Input matrix (as linearized array of values)
+    MutableArrayRef<Value*> submatrix,      // Output matrix (ditto)
+    uint32_t                size,           // Size of input matrix
+    uint32_t                rowToDelete,    // Row index to delete
+    uint32_t                columnToDelete) // Column index to delete
+{
+    uint32_t inIdx = 0, outIdx = 0;
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: inIdx -> inElementIdx, outIdx -> outElementIdx

File: builder/llpcBuilderImplMatrix.cpp [hunk 12]
```diff
@@ -203,5 +209,180 @@ Value* BuilderImplMatrix::CreateOuterProduct(
...
+    // where Adj(a) is the cofactor of a, which is the determinant of the submatrix obtained by deleting
+    // the row and column of a, then negated if row+col is odd.
+
+    SmallVector<Value*, 16> resultElements;
+    resultElements.resize(size * size);
+    SmallVector<Value*, 9> submatrix;
+    submatrix.resize((size - 1) * (size - 1));
+
+    // Calculate reciprocal of determinant, and negated reciprocal of determinant.
+    Value* pRecipDet = CreateFDiv(ConstantFP::get(elements[0]->getType(), 1.0), Determinant(elements, size));
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: pRecipDet -> pRcpDet

File: builder/llpcBuilderImplMatrix.cpp [hunk 13]
```diff
@@ -203,5 +209,180 @@ Value* BuilderImplMatrix::CreateOuterProduct(
...
+    Value* pNegRecipDet = CreateFSub(Constant::getNullValue(elements[0]->getType()), pRecipDet);
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: pNegRecipDet -> pNegRcpDet

File: builder/llpcBuilderImplMatrix.cpp [hunk 14]
```diff
@@ -203,5 +209,180 @@ Value* BuilderImplMatrix::CreateOuterProduct(
...
+    }
+    return pResult;
+}
+
+// =====================================================================================================================
+// Get submatrix by deleting specified row and column
+void BuilderImplMatrix::GetSubmatrix(
+    ArrayRef<Value*>        matrix,         // Input matrix (as linearized array of values)
+    MutableArrayRef<Value*> submatrix,      // Output matrix (ditto)
+    uint32_t                size,           // Size of input matrix
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: size -> order // Order of input matrix

File: builder/llpcBuilderImpl.h [hunk 15]
```diff
@@ -472,9 +472,25 @@ class BuilderImplMatrix : virtual public BuilderImplBase
...
+
+    // Get submatrix by deleting specified row and column
+    void GetSubmatrix(ArrayRef<Value*>        matrix,
+                      MutableArrayRef<Value*> submatrix,
+                      uint32_t                size,
```
Comment (amdrexu) [inline builder/llpcBuilderImpl.h]: size -> order

**Comment by trenouf** [comment]
V2: Addressed review comments.

**Comment by trenouf** [comment]
Thanks Rex. I will land this if it passes testing.

**Comment by amdrexu** [comment]
I think you are going to add ALU instructions to builder soon. Please make a BuilderImplArith class for those ALU operations. Please do not make a large change to incorporate all and follow the order of this:
native operator, GLSL angle and trigonometry functions, exponential fucntions, common ALU functions, pack/unpack functions, geometric functions, integer functions.

You can reference glslArithOpEmu.ll and add comments for each subpart.

We still have three .ll, check if we can remove them: glslInlineConstOpEmu.ll, glslPushConstOpEmu.ll, glslSharedVarOpEmu.ll in the future.

**Comment by trenouf** [comment]
Yes, I will be adding BuilderImplArith. I will probably be doing them in a different order; trig and exponential will probably be the first to be ready to review. I have been referencing glslArithOpEmu.ll.

**Comment by amdrexu** [comment]
Any order is OK as long as we group related ones to categories.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 13b12ab5fafa93438b31722c9aa2b82bd338385d
<details open><summary>Driver commits used in build</summary><ul>

<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 66e78b997748d03d77e1d706c10f1f17e18e5654</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 9b632ef4f132bddc94769702ed8b49efbc39d89c</li>
<li>LLVM: amd-vulkan-dev 9bc5dd4450a6361faf5c5661056a7ee494fad830</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 282837/282837 (100.0%)</li>
<li>Failed: 0/282837 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366376/366376 (100.0%)</li>
<li>Failed: 0/366376 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**